### PR TITLE
Andriy

### PR DESCRIPTION
**Pull request by Pasemko**

### CODE REVIEW

**Review by BogdanKalysh — APPROVED** [review]

**Review by VolodymyrVoloshyn02 — APPROVED** [review]

**Review by Nazar-TK — APPROVED** [review]

File: DAL/headers/constants.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+namespace validationMessage {
+
+}
+
+namespace dal {
+
+
+    const QString DBTYPE = "QPSQL";
```
Comment (AndriyFedorenko) [inline DAL/headers/constants.h:30]: It is ok as a temporary solution but it should be fixed in the future.

File: DAL/headers/repositories/userrepository.h [hunk 2]
```diff
@@ -4,6 +4,7 @@
 #include "irepository.h"
 #include "user.h"
 
+
```
Comment (AndriyFedorenko) [inline DAL/headers/repositories/userrepository.h:7]: remove empty line

File: DAL/sources/repositories/categoryrepository.cpp [hunk 3]
```diff
@@ -0,0 +1,105 @@
+#include "categoryrepository.h"
+#include "querybuilders.h"
+#include "constants.h"
+
+using namespace dal;
+QVector<Category> CategoryRepository::select(QString query)
```
Comment (AndriyFedorenko) [inline DAL/sources/repositories/categoryrepository.cpp:6]: Parameter query should be passed as const reference

File: DAL/sources/repositories/categoryrepository.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            categories.push_back(Category(
+                                result.value(ID).toInt(),
+                                result.value(NAME).toString(),
+                                (Type)result.value(TYPE).toInt(),
+                                QColor(result.value(COLOR).toString()),
+                                result.value(CATEGORY_ID).toInt()
+                                ));
+        }
+    }
+    else
```
Comment (AndriyFedorenko) [inline DAL/sources/repositories/categoryrepository.cpp:28]: Please add curly braces

File: DAL/sources/repositories/categoryrepository.cpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    categories.push_back(Category(1, "Credits", Type::expense, QColor("#d41111"), 1));
+    return categories;
+}
+
+bool CategoryRepository::update(Category &object)
+{
+    QSqlDatabase db = setUpDatabase();
+
+    if(db.open())
+    {
```
Comment (AndriyFedorenko) [inline DAL/sources/repositories/categoryrepository.cpp:39]: You are using different styles. 
You should decide if you are using 
if () { 
or
 if ()
{

File: DAL/sources/repositories/categoryrepository.cpp [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        QSqlQuery query;
+        QVector<QString> fields;
+        fields << NAME << TYPE << COLOR;
+        QVector<QString> values;
+        QString type;
+        if(object.getType() == 0)
+            type = "income";
```
Comment (AndriyFedorenko) [inline DAL/sources/repositories/categoryrepository.cpp:46]: Please move this text to constants.

File: DAL/sources/repositories/categoryrepository.cpp [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    QSqlDatabase db = setUpDatabase();
+
+    if(db.open())
+    {
+        QSqlQuery query;
+        QVector<QString> fields;
+        fields << NAME << TYPE << COLOR;
+        QVector<QString> values;
+        QString type;
+        if(object.getType() == 0)
```
Comment (AndriyFedorenko) [inline DAL/sources/repositories/categoryrepository.cpp:45]: better to use operator ? :

File: DAL/sources/repositories/querybuilders.cpp [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+
+QString updateQueryBuilder(QPair<QString, int> tableCredentials, QVector<QString> fields, QVector<QString> values)
+{
+    //      EXAMPLE QUERY
+    //  UPDATE table_name
+    //  SET column1 = value1, column2 = value2, ...
+    //  WHERE condition
+    if(fields.size() == values.size()) {
```
Comment (AndriyFedorenko) [inline DAL/sources/repositories/querybuilders.cpp:36]: The same

File: DAL/sources/repositories/transactionrepository.cpp [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                                result.value(CATEGORY_ID).toInt()
+                                ));
+        }
+    }
+    else
+        qDebug() << db.lastError();
+    return transactions;
+}
+
+bool TransactionRepository::update(Transaction &object)
```
Comment (AndriyFedorenko) [inline DAL/sources/repositories/transactionrepository.cpp:29]: You are not changing object. Please use const

**Review by AndriyFedorenko — CHANGES_REQUESTED** [review]

File: DAL/headers/models/user.h [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+public:
+    User(int id, QString name, QString email, QString password, int balance);
+    int getId() const;
+    QString getName() const;
+    QString getPassword() const;
+    QString getEmail() const;
+    int getBalance() const;
+
+    QString setId(const int id);
+    QString setName(const QString name);
```
Comment (AndriyFedorenko) [inline DAL/headers/models/user.h:23]: Please add reference

File: DAL/sources/models/category.cpp [hunk 11]
```diff
@@ -7,14 +7,16 @@ Category::Category()
     this->name = QString();
     this->type = income;
     this->color = QColor();
+    this->userId = -1;
 }
 
-Category::Category(int id, QString name, Type type, QColor color)
+Category::Category(const int id, const QString name, const Type type, const QColor color, const int userId)
```
Comment (AndriyFedorenko) [inline DAL/sources/models/category.cpp:13]: Please add reference

File: DAL/sources/models/category.cpp [hunk 12]
```diff
@@ -25,50 +27,61 @@ Category::Category(const Category &other)
...
-QString Category::setId(int id)
+QString Category::setId(const int id)
 {
     // TODO: Int validation logic
     this->id = id;
     return "Success";
 }
 
-QString Category::setName(QString name)
+QString Category::setName(const QString name)
```
Comment (AndriyFedorenko) [inline DAL/sources/models/category.cpp:62]: Please add reference

File: DAL/sources/models/transaction.cpp [hunk 13]
```diff
@@ -20,64 +20,57 @@ Transaction::Transaction(int id, int amount, QDate date, QString description, in
...
 {
     return this->description;
 }
 
-QString Transaction::setId(int id)
+QString Transaction::setId(const int id)
 {
     // TODO: Int validation logic
     this->id = id;
     return "Success";
```
Comment (AndriyFedorenko) [inline DAL/sources/models/transaction.cpp:47]: it is more effective to use an enumeration for a returning value

File: DAL/sources/models/transaction.cpp [hunk 14]
```diff
@@ -20,64 +20,57 @@ Transaction::Transaction(int id, int amount, QDate date, QString description, in
...
-QString Transaction::setDate(QDate date)
+QString Transaction::setDate(const QDate date)
 {
     // TODO: Validation logic; It may be realized with the help of QDate::isValid
     this->date = date;
     return "Success";
 }
 
-QString Transaction::setDescription(QString description)
+QString Transaction::setDescription(const QString description)
```
Comment (AndriyFedorenko) [inline DAL/sources/models/transaction.cpp:64]: Please add a reference for all QString parameters

File: DAL/sources/repositories/userrepository.cpp [hunk 15]
```diff
@@ -0,0 +1,105 @@
+#include "userrepository.h"
+
+using namespace dal;
+
+QVector<User> UserRepository::select(const QString &query)
+{
+
```
Comment (AndriyFedorenko) [inline DAL/sources/repositories/userrepository.cpp:7]: remove empty line

**Review by AndriyFedorenko — APPROVED** [review]
Generally looks good but please add a reference for all QString parameters (This is the rule for all complicated types)

**Pull request merged**